src/kdf.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/kdf.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw {
+
+KeyDerivationFunction::~KeyDerivationFunction() = default;
+
+
+
+
+class PBKDF2::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction, uint32_t iterations):
+        _hashFunction(hashFunction), _iterations(iterations) {}
+
+    ~Impl() = default;
+
+    Impl(Impl&&) = default;
+
+    Impl(Impl&) = default;
+
+    std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password,
+                                   const size_t outputLength,
+                                   const std::vector<uint8_t> &salt)
+    {
+        std::vector<uint8_t> derivedKey;
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(_hashFunction);
+        derivedKey.resize(outputLength);
```

Comment:
```suggestion
        std::vector<uint8_t> derivedKey(outputLength);
        const EVP_MD* digestFn = _getMDPtrFromDigestType(_hashFunction);
```

This saves a realloc.